File: mq_consumer.py
import os
import time
import json
import pymqi

queue_manager = os.getenv("MQ_QMGR_NAME", "AUTORIZA")
channel = os.getenv("MQ_CHANNEL", "SYSTEM.ADMIN.SVRCONN")
host = os.getenv("MQ_HOST", "ibmmq")
port = os.getenv("MQ_PORT", "1414")
queue_name = os.getenv("MQ_QUEUE", "BOFTD_ENV")
user = os.getenv("MQ_USER", "admin")
password = os.getenv("MQ_PASSWORD", "admin123")

conn_info = f"{host}({port})"

# Esperar a que MQ esté disponible con reintentos
qmgr = None
attempt = 0
while qmgr is None:
    try:
        attempt += 1
        print(f"Conectando a MQ {queue_manager} en {conn_info} (intento {attempt})...")
        qmgr = pymqi.connect(queue_manager, channel, conn_info, user, password)
        print("Conexión a MQ exitosa.")
    except Exception as e:
        print(f"Conexión fallida: {e}. Reintentando en 3s...")
        time.sleep(3)

queue_open_opts = pymqi.CMQC.MQOO_INPUT_AS_Q_DEF | pymqi.CMQC.MQOO_FAIL_IF_QUIESCING
queue = pymqi.Queue(qmgr, queue_name, queue_open_opts)

md = pymqi.MD()
gmo = pymqi.GMO()
gmo.Options = (
    pymqi.CMQC.MQGMO_WAIT
    | pymqi.CMQC.MQGMO_NO_SYNCPOINT
    | pymqi.CMQC.MQGMO_FAIL_IF_QUIESCING
)
gmo.WaitInterval = 2000  # 2 segundos

print(f"Escuchando cola '{queue_name}' con espera de {gmo.WaitInterval}ms...")

try:
    while True:
        try:
            # Lee hasta 1MB por mensaje
            msg = queue.get(1024 * 1024, md, gmo)
        except pymqi.MQMIError as e:
            if e.comp == 2 and e.reason == pymqi.CMQC.MQRC_NO_MSG_AVAILABLE:
                continue
            raise

        try:
            text = msg.decode("utf-8")
        except Exception:
            text = str(msg)

        # Intentar parsear JSON para mostrarlo bonito
        try:
            obj = json.loads(text)
            resumen = obj.get("id_lote") if isinstance(obj, dict) else None
            if resumen is not None:
                print(f"Recibido id_lote={resumen}")
            else:
                print(f"Recibido JSON: {obj}")
        except Exception:
            print(f"Recibido (texto): {text}")

except KeyboardInterrupt:
    print("Interrumpido por el usuario.")
finally:
    try:
        queue.close()
    finally:
        if qmgr is not None:
            qmgr.disconnect()

File: mq_consumer.php
<?php
// Simple PHP consumer for IBM MQ using the mqseries extension (PECL)

$queueManager = getenv('MQ_QMGR_NAME') ?: 'AUTORIZA';
$channel      = getenv('MQ_CHANNEL')    ?: 'DEV.APP.SVRCONN';
$host         = getenv('MQ_HOST')       ?: 'ibmmq';
$port         = getenv('MQ_PORT')       ?: '1414';
$queueName    = getenv('MQ_QUEUE')      ?: 'BOFTD_ENV';
$user         = getenv('MQ_USER')       ?: 'app';
$password     = getenv('MQ_PASSWORD')   ?: 'passw0rd';

$connName = $host . '(' . $port . ')';

// Build Client Channel Definition (CD)
$cd = [
    'ChannelName'    => $channel,
    'ConnectionName' => $connName,
    'TransportType'  => MQSERIES_MQXPT_TCP,
];

// Security parms for userid/password
$sp = [
    'UserId'   => $user,
    'Password' => $password,
];

// Connection options (CNO) with client connection and security parms
$cno = [
    'Options'       => MQSERIES_MQCNO_HANDLE_SHARE_NO_BLOCK,
    'ClientConn'    => $cd,
    'SecurityParms' => $sp,
];

$hConn = null; $compCode = 0; $reason = 0;
mqseries_connx($queueManager, $cno, $hConn, $compCode, $reason);
if ($compCode !== MQSERIES_MQCC_OK) {
    fwrite(STDERR, "MQ CONNX failed: CC=$compCode RC=$reason\n");
    exit(1);
}
fwrite(STDOUT, "Connected to MQ $queueManager via $channel @ $connName\n");

// Open queue for input
$od = [ 'ObjectName' => $queueName, 'ObjectQMgrName' => $queueManager ];
$openOpts = MQSERIES_MQOO_INPUT_AS_Q_DEF | MQSERIES_MQOO_FAIL_IF_QUIESCING;
$hObj = null; $compCode = 0; $reason = 0;
mqseries_open($hConn, $od, $openOpts, $hObj, $compCode, $reason);
if ($compCode !== MQSERIES_MQCC_OK) {
    fwrite(STDERR, "MQ OPEN failed: CC=$compCode RC=$reason\n");
    mqseries_disc($hConn, $compCode, $reason);
    exit(1);
}
fwrite(STDOUT, "Opened queue '$queueName' for reading\n");

// Get message options: wait up to 2 seconds, convert to local codepage
$md  = [];
$gmo = [
    'Options'      => MQSERIES_MQGMO_WAIT | MQSERIES_MQGMO_NO_SYNCPOINT | MQSERIES_MQGMO_CONVERT | MQSERIES_MQGMO_FAIL_IF_QUIESCING,
    'WaitInterval' => 2000,
];

// Read loop
while (true) {
    $bufferLen = 1024 * 1024; // 1 MB
    $msg = '';
    $dataLen = 0;
    $compCode = 0; $reason = 0;
    mqseries_get($hConn, $hObj, $md, $gmo, $bufferLen, $msg, $dataLen, $compCode, $reason);

    if ($compCode === MQSERIES_MQCC_FAILED && $reason === MQSERIES_MQRC_NO_MSG_AVAILABLE) {
        // Timeout waiting for message
        continue;
    }
    if ($compCode !== MQSERIES_MQCC_OK) {
        fwrite(STDERR, "MQ GET failed: CC=$compCode RC=$reason\n");
        break;
    }

    // Try JSON decode for nicer output
    $text = $msg; // already converted
    $decoded = json_decode($text, true);
    if (json_last_error() === JSON_ERROR_NONE) {
        if (is_array($decoded) && array_key_exists('id_lote', $decoded)) {
            echo "Recibido id_lote=" . $decoded['id_lote'] . "\n";
        } else {
            echo "Recibido JSON: " . json_encode($decoded) . "\n";
        }
    } else {
        echo "Recibido (texto): $text\n";
    }
}

// Cleanup
mqseries_close($hConn, $hObj, MQSERIES_MQCO_NONE, $compCode, $reason);
mqseries_disc($hConn, $compCode, $reason);
?>

